bart/sched/pelt.py

```diff
@@ -0,0 +1,492 @@
+        a task starts to run we consider it as if if was running for the
+        entire previous PELT interval. To fix this, by exploting the knowledge
+        that a task is always configured using an integer number of run/speep
+        PELT samples, we can look at the 1us before time wrt the specified
+        sample time.
+        """
+        # The 1us "before time" is always represeting the exact status of the
+        # task in that period. This is why we remove 1 in the following
+        # computation:
+        period_delta_us = (float(time_ms) * 1e3 - 1) % self.period_us
+        if period_delta_us <= self.run_us:
+            return True
+        return False
+
+    def __str__(self):
+        return "PeriodicTask(start: {:.3f} [ms], period: {:.3f} [ms], run: {:.3f} [ms], "\
+               "duty_cycle: {:.3f} [%], pelt_avg: {:d})"\
+                .format(self.start_us/1e3, self.period_us/1e3,
+                        self.run_us/1e3, self.duty_cycle_pct,
+                        self.pelt_avg)
+
+
+################################################################################
+# PELT Model
+################################################################################
+
+_PELTStats = namedtuple('PELTStats', [
+    'start_s', 'end_s',
+    'pelt_avg', 'pelt_init', 'half_life',
+    'tmin', 'tmax', 'min', 'max', 'avg', 'std',
+    'err', 'err_pct'])
+class PELTStats(_PELTStats):
+    """Statistic of a computed PELT signal
+
+    This is a collection of metrics computed on a simulated PELT signal:
+    start_s   : start time [s] used for PELT signal simulation
+    end_s     : end time [s] used for PELT signal simulation
+    pelt_avg  : expected average value of the PELT signal
+    pelt_init : initial PELT value
+    halt_life : time [ms] of the PELT's halt-life parameter
+    tmin      : start time [s] used for statistics computation
+    tmax      : end time [s] used for statistics computation
+    min       : minimum PELT value, within [tmin, tmax]
+    max       : maximum PELT value, within [tmin, tmax]
+    avg       : average of the PELT values, within [tmin, tmax]
+    std       : standard deviation of the PELT value, within [tmin, tmax]
+    err       : difference between pelt_avg and avg
+    err_pct   : the avg error percentage, compared to the expected average (pelt_avg)
+
+    :note: the [start_s, end_s] interval used for statistical computations can be
+           smaller than the timespan of the simulated PETL signal. For example, we
+           can simulate a PELT signal for 3 [s] but compute statistics only for the
+           last 1 [s]. This is allows for example to ignore the initial execution
+           of a task while focusing for stats only on a timeframe in which the
+           signal can be considered stable.
+    """
+    pass
+
+_PELTRange = namedtuple('PELTRange', [
+    'min_value', 'max_value'])
+class PELTRange(_PELTRange):
```

Comment:
Love the inheritance trick :tongue: 

I think you could even avoid the `_PELTRange` name by putting the `namedtuple` call inside the brackets. (Not asking you to... just interesting)